### PR TITLE
resolve inconsistencies between bufr2ioda python converters and jcb templates

### LINKED ISSUE #1227: Inconsistency between bufr2ioda iodafile and init obsdatain
The following bufr2ioda converters add the string `data_format` to the name of the iodafile created by the converter
```
./bufr2ioda_sfcshp_prepbufr.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}.{data_format}.nc"
./bufr2ioda_snocvr_bufr.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}.{data_format}.nc"
./bufr2ioda_gpsro_bufr_combined.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}.{data_format}.nc"
./bufr2ioda_acft_profiles_prepbufr.py:    iodafile = f"{cycle_type}.t{hh}z.acft_profiles.{data_format}.nc"
./marine/bufr2ioda_insitu_surface_altkob.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}_profiles.{data_format}.nc"
./marine/bufr2ioda_insitu_profile_marinemammal.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}_profiles.{data_format}.nc"
./marine/bufr2ioda_insitu_surface_trkob.py:    iodafile = f"{cycle_type}.t{hh}z.insitu_surface_{data_format}.{cycle_datetime}.nc4"
./marine/bufr2ioda_insitu_profile_bathy.py:    iodafile = f"{cycle_type}.t{hh}z.insitu_profile_{data_format}.{cycle_datetime}.nc4"
./marine/bufr2ioda_insitu_profile_xbtctd.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}_profiles.{data_format}.nc4"
./marine/bufr2ioda_insitu_profile_tesac.py:    iodafile = f"{cycle_type}.t{hh}z.insitu_profile_{data_format}.{cycle_datetime}.nc4"
./marine/bkp/bufr2ioda_bathythermal_profiles.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}_profiles.{data_format}.nc"
./marine/bkp/bufr2ioda_trackob_surface.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}_profiles.{data_format}.nc"
./marine/bkp/bufr2ioda_subpfl_argo_profiles.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}_profiles.{data_format}.nc"
./marine/bkp/bufr2ioda_xbtctd_profiles.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}_profiles.{data_format}.nc"
./marine/bkp/bufr2ioda_altkob_surface.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}_profiles.{data_format}.nc"
./marine/bkp/bufr2ioda_tesac_mammals_profiles.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}_profiles.{data_format}.nc"
./marine/bkp/bufr2ioda_tesac_profiles.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}_profiles.{data_format}.nc"
./marine/bkp/bufr2ioda_subpfl_glider_profiles.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}_profiles.{data_format}.nc"
./marine/bufr2ioda_insitu_profile_glider.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}_profiles.{data_format}.nc4"
./bufr2ioda_adpupa_prepbufr.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}.{data_format}.nc"
./bufr2ioda_gnssro_bufr.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}.tm00.{data_format}.nc"
./bufr2ioda_adpsfc_prepbufr.py:    iodafile = f"{cycle_type}.t{hh}z.{data_type}.{data_format}.nc"
./bufr2ioda_conventional_prepbufr_ps.py:    iodafile = f"{cycle_type}.t{hh}z.{data_description}.{data_format}.nc"
```
Inclusion of `data_format` in the ioda filename is at odds with the obsdatin filename written to the var and local ensemble yamls by the init jobs.   The obsdatain filenames do not include the `data_format` string.  This inconsistency causes the var and local ensemble init jobs to abort because an expected ioda observation file is not found.

For example, `./bufr2ioda_gnssro_bufr.py` creates ioda file `gdas.t18z.gnssro.bufr_d.nc`.  However, the var yaml expects to read `gdas.t18z.gnssro.tm00.nc`.  The init jobs does not find this file and aborts  
```
OSError: unable to copy /scratch1/NCEPDEV/stmp2/role.jedipara/COMROOT/prjedi/gdas.20240502/18/obs/gdas.t18z.gnssro.tm00.nc to /scratch1/NCEPDEV/stmp2/role.j\
edipara/RUNDIRS/prjedi/gdasatmanl_18/obs//gdas.t18z.gnssro.tm00.nc
+ JGLOBAL_ATM_ANALYSIS_INITIALIZE[1]: postamble JGLOBAL_ATM_ANALYSIS_INITIALIZE 1721392162 1
```

### PR DESCRIPTION
**Pull request by RussTreadon-NOAA**
bufr2ioda python converters use different naming conventions for output ioda files.  JCB templates assume certain patterns.  This PR updates the output ioda  filenames in bufr2ioda python converters to be consistent with JCB templates.  

This PR focuses on bufr2ioda python scripts for select observation types used in atmospheric variational and local ensemble DA.  Marine DA converters are not altered in this PR.

Resolves #1227

### CODE REVIEW

**Comment by RussTreadon-NOAA** [comment]
Convert to draft until in office to actively test.

**Comment by emilyhcliu** [comment]
> Convert to draft until in office to actively test.

@RussTreadon-NOAA I can test this and report the test results here.  
I will do this during JEDI Junction time tomorrow.

**Comment by RussTreadon-NOAA** [comment]
@emilyhcliu , thank you.  I have already tested most of the modified python bufr2ioda converters in this PR via global-workflow  job prepatmiodaobs.  The generated ioda files have been processed by the g-w JEDI ATM variational and local ensemble DA apps.  

There is no science in this PR.  It simply aligns atmospheric bufr2ioda output filenames with assumed patterns in JCB templates.   No changes are made to marine bufr2ioda python converter scripts (ie, those in `/ush/ioda/bufr2ioda/marine`).

**Comment by RussTreadon-NOAA** [comment]
Mark PR as _Ready for review_ so @emilyhcliu can comment & review.

File: ush/ioda/bufr2ioda/bufr2ioda_conventional_prepbufr_ps.py
Comment (emilyhcliu) [inline ush/ioda/bufr2ioda/bufr2ioda_conventional_prepbufr_ps.py:867]: `observation_from_jcb` for station pressure is defined as `conv_ps` (in [jcb-prototype_3dvar.yam.j2)](https://github.com/NOAA-EMC/GDASApp/blob/develop/parm/atm/jcb-prototype_3dvar.yaml.j2)

So,  I changed the `data_description` in ioda-conventional_prepbufr_ps.json:
conventional_ps ---> conv_ps

I ran the 3dvar with conv_ps only.  The `gdasatmanlvar` ran to completion with results closely matched GSI results.

Comment (RussTreadon-NOAA) [reply ush/ioda/bufr2ioda/bufr2ioda_conventional_prepbufr_ps.py:867]: Thank you @emilyhcliu for this test.  

I made a different change in my local working copy.  I replaced `conv_ps` in `parm/atm/jcb-prototype_3dvar.yaml.j2` with `conventional_ps`.  The same change was made to `parm/atm/jcb-prototype_lgetkf.yaml.j2`

Which approach do we adopt moving forward?  

Do we 
1. edit the `observations:` lists in `parm/atm` yamls, or
2. edit the `json` files in `parm/ioda/bufr2ioda`?

Comment (emilyhcliu) [reply ush/ioda/bufr2ioda/bufr2ioda_conventional_prepbufr_ps.py:867]: @RussTreadon-NOAA Both ways work.  Let's go with your change logic:
1, add tm00 to the ioda obs filename in the Python converter code. 
2. edit observations: lists in param/atm
3. rename conv_ps.yaml.j2 to conventional_ps.yaml.j2 (under parm/jcb-gdas/observations/atmosphere)

Comment (RussTreadon-NOAA) [reply ush/ioda/bufr2ioda/bufr2ioda_conventional_prepbufr_ps.py:867]: Another change is required with my approach.  

We need to move `parm/jcb-gdas/observations/atmosphere/conv_ps.yaml.j2` to `parm/jcb-gdas/observations/atmosphere/conventional_ps.yaml.j2`.  Without this change, `atmanlinit` fails with 
```
jinja2.exceptions.TemplateNotFound: conventional_ps.yaml.j2
```

**Comment by RussTreadon-NOAA** [comment]
@YoulongXia-NOAA and @jiaruidong2017 , `ush/ioda/bufr2ioda/bufr2ioda_snocvr_bufr.py` defines `iodafile` as
```
    iodafile = f"{cycle_type}.t{hh}z.{data_type}.nc"
```
I see that `./snow/jcb-base.yaml.j2` contains
```
atm_obsdatain_suffix: ".tm00.nc"
```
Given this, it seems we need to replace `data_type` in the `iodafile` line above with
```
    iodafile = f"{cycle_type}.t{hh}z.{data_type}.tm00.nc"
```

Do you agree?

**Comment by YoulongXia-NOAA** [comment]
@RussTreadon-NOAA, I am fine for a test as I just develop ioda-converters for snowcvr data. Let Jiarui @jiaruidong2017 to reply your question more precisely  as he is working on a combined snow dataset and his work is more associated with GDASApp and global workflow tasks. Thank you.

**Comment by jiaruidong2017** [comment]
@RussTreadon-NOAA I am fine with any changes to the filename format as long as it keeps consistent.  

When I check the `iodafile` in the `ush/ioda/bufr2ioda/bufr2ioda_snocvr_bufr.py`, the current form is `iodafile = f"{cycle_type}.t{hh}z.{data_type}.{data_format}.nc"` with `.{data_format}` there.

**Comment by jiaruidong2017** [comment]
@RussTreadon-NOAA The `bufrfile` in `ush/ioda/bufr2ioda/bufr2ioda_snocvr_bufr.py` is defined as below: 

`bufrfile = f"{cycle_type}.t{hh}z.{data_type}.tm00.{data_format}"
`
including `tm00`.

**Comment by jiaruidong2017** [comment]
> @RussTreadon-NOAA I am fine with any changes to the filename format as long as it keeps consistent.
> 
> When I check the `iodafile` in the `ush/ioda/bufr2ioda/bufr2ioda_snocvr_bufr.py`, the current form is `iodafile = f"{cycle_type}.t{hh}z.{data_type}.{data_format}.nc"` with `.{data_format}` there.

@RussTreadon-NOAA The `iodafile` need to change from:

`iodafile = f"{cycle_type}.t{hh}z.{data_type}.{data_format}.nc"`

TO:

`iodafile = f"{cycle_type}.t{hh}z.{data_type}.{data_format}"`

**Comment by RussTreadon-NOAA** [comment]
This is interesting, @jiaruidong2017 . 

File `parm/snow/jcb-prototype_3dvar.yaml.j2` contains
```
# Naming conventions for observational files
atm_obsdatain_path: "{{atm_obsdatain_path}}"
atm_obsdatain_prefix: "{{OPREFIX}}"
atm_obsdatain_suffix: ".tm00.nc"
```
This suggests that the suffix should be `.tm00.nc`.   

Given your input I will not modify `ush/ioda/bufr2ioda/bufr2ioda_snocvr_bufr.py` in this PR.  If needed, you can update this converter in a separate PR.

**Comment by jiaruidong2017** [comment]
Thanks @RussTreadon-NOAA We didn't use `bufr2ioda_snocvr_bufr.py` yet to convert SNOCVR BUFR data into ioda file for the CI test. You can make the changes in your PR, and we can follow the changes to define the ioda file later.

**Comment by RussTreadon-NOAA** [comment]
@emilyhcliu , for the 2024050218 gdas, the prepatmiodaobs job creates the following ioda format dump files
```
gdas.t18z.ADPSFC.tm00.nc
gdas.t18z.ADPUPA.tm00.nc
gdas.t18z.SFCSHP.tm00.nc
gdas.t18z.acft_profiles.tm00.nc
gdas.t18z.ascatw.ascat_metop-b.tm00.nc
gdas.t18z.ascatw.ascat_metop-c.tm00.nc
gdas.t18z.atms_n20.tm00.nc
gdas.t18z.atms_npp.tm00.nc
gdas.t18z.conventional_ps.tm00.nc
gdas.t18z.gnssro.tm00.nc
gdas.t18z.gpsro.tm00.nc
gdas.t18z.mtiasi_metop-b.tm00.nc
gdas.t18z.mtiasi_metop-c.tm00.nc
gdas.t18z.ompsnp_n20.tm00.nc
gdas.t18z.ompsnp_npp.tm00.nc
gdas.t18z.ompstc_n20.tm00.nc
gdas.t18z.ompstc_npp.tm00.nc
gdas.t18z.satwnd.abi_goes-16.tm00.nc
gdas.t18z.satwnd.abi_goes-18.tm00.nc
gdas.t18z.satwnd.leogeo_multi.tm00.nc
gdas.t18z.satwnd.viirs_n20.tm00.nc
gdas.t18z.satwnd.viirs_npp.tm00.nc
```
A few questions:
1. We don't have yamls for all these dump files in [`parm/jcb-gdas/observations/atmosphere`](https://github.com/NOAA-EMC/jcb-gdas/tree/9c322d98d7aab4e1cfca8588f6913d8bee190dde/observations/atmosphere).  Should we have an observation yaml for each ioda format dump file?   
2. Do we have a problem with inconsistency between the names of the ioda format dump files and the names of the observation yamls?  For example, we have ioda format dump file `SFCSHP` and observation yaml `sfcship`.  Do these two go together?
3. We have yamls in [`parm/jcb-gdas/observations/atmosphere`](https://github.com/NOAA-EMC/jcb-gdas/tree/9c322d98d7aab4e1cfca8588f6913d8bee190dde/observations/atmosphere) for which we do not have bufr2ioda converters in `ush/ioda/bufr2ioda`.  Examples include `cris-fsr-n20.yaml.j2` and `ssmis_f17.yaml.j2`.  Is there no buf2ioda converter script for these observation types because the plan to directly read the bufr dump file via the ioda backend?

**Comment by RussTreadon-NOAA** [comment]
> Thanks @RussTreadon-NOAA We didn't use `bufr2ioda_snocvr_bufr.py` yet to convert SNOCVR BUFR data into ioda file for the CI test. You can make the changes in your PR, and we can follow the changes to define the ioda file later.

I am reluctant to change things I can not test.  I will not modify `bufr2ioda_snocvr_bufr.py` in this PR.

**Comment by emilyhcliu** [comment]
@RussTreadon-NOAA. I break down the list of observations from you into two lists (A and B)

Here is the list of observations - List-A
These observations were validated in the end-to-end test before the implementation of JCB.  
```
gdas.t18z.ascatw.ascat_metop-b.tm00.nc
gdas.t18z.ascatw.ascat_metop-c.tm00.nc
gdas.t18z.mtiasi_metop-b.tm00.nc
gdas.t18z.mtiasi_metop-c.tm00.nc
gdas.t18z.atms_n20.tm00.nc
gdas.t18z.atms_npp.tm00.nc
gdas.t18z.conventional_ps.tm00.nc
gdas.t18z.gnssro.tm00.nc
gdas.t18z.gpsro.tm00.nc
gdas.t18z.ompsnp_n20.tm00.nc
gdas.t18z.ompsnp_npp.tm00.nc
gdas.t18z.ompstc_n20.tm00.nc
gdas.t18z.ompstc_npp.tm00.nc
gdas.t18z.satwnd.abi_goes-16.tm00.nc
gdas.t18z.satwnd.abi_goes-18.tm00.nc
gdas.t18z.satwnd.leogeo_multi.tm00.nc
gdas.t18z.satwnd.viirs_n20.tm00.nc
gdas.t18z.satwnd.viirs_npp.tm00.nc
```

Here is the list of observations  - List-B 
These are conventional data.  Andrew is leading the evaluation of these observation types.
```
gdas.t18z.ADPSFC.tm00.nc
gdas.t18z.ADPUPA.tm00.nc
gdas.t18z.SFCSHP.tm00.nc
gdas.t18z.acft_profiles.tm00.nc
```
So, Let's deal with the observations in List-A first.
For observation in List-A, the conventional_ps is tested with JCB and will be modified to your way in this PR. 
For the other observation in List-A, I opened [Issue #1240](https://github.com/NOAA-EMC/GDASApp/issues/1240).
I will check one by one to make sure they work properly with JCB and update filter options such as `reduce obs space`.
With the reduced obs space, cycling DA is more feasible (e.g., for IASI, only data after thinning will be kept in the obs space for variational analysis).  

Here are slides regarding `reduce obs space` I shared last Thursday during EMC JEDI Junction

20240731 JEDI T2O
20240731 JEDI T2O (3)

Am I heading in the right direction?

**Comment by RussTreadon-NOAA** [comment]
Thank you @emilyhcliu for the update and the work you are doing in issue #1240.

I'll continue this PR by focusing on what I can get through prepatmiodaobs, atmanlinit, and atmanlvar via minor script and filename changes.  I will probably hold off on local ensemble DA until g-w issue [#2415](https://github.com/NOAA-EMC/global-workflow/issues/2415) has a PR.

**Comment by RussTreadon-NOAA** [comment]
**Hercules ctests**
Install `RussTreadon-NOAA:feature/bufr2ioda` at [`182086d`](https://github.com/RussTreadon-NOAA/GDASApp/commit/182086db235257155f6b94b3cd69272ea4dc5dfe) on Hercules in g-w `develop` at [`e2c0f06`](https://github.com/NOAA-EMC/global-workflow/commit/e2c0f0671a59628f4614637c01900cb265d837d5).  Run ctests with following results
```
Test project /work/noaa/da/rtreadon/git/global-workflow/test_hercules/sorc/gdas.cd/build
      Start 1532: test_gdasapp_util_coding_norms
 1/44 Test #1532: test_gdasapp_util_coding_norms ........................   Passed    4.68 sec
      Start 1533: test_gdasapp_util_ioda_example
 2/44 Test #1533: test_gdasapp_util_ioda_example ........................   Passed   11.80 sec
      Start 1534: test_gdasapp_util_prepdata
 3/44 Test #1534: test_gdasapp_util_prepdata ............................   Passed    4.80 sec
      Start 1535: test_gdasapp_util_rads2ioda
 4/44 Test #1535: test_gdasapp_util_rads2ioda ...........................   Passed    0.48 sec
      Start 1536: test_gdasapp_util_ghrsst2ioda
 5/44 Test #1536: test_gdasapp_util_ghrsst2ioda .........................   Passed    0.11 sec
      Start 1537: test_gdasapp_util_rtofstmp
 6/44 Test #1537: test_gdasapp_util_rtofstmp ............................   Passed    1.83 sec
      Start 1538: test_gdasapp_util_rtofssal
 7/44 Test #1538: test_gdasapp_util_rtofssal ............................   Passed    0.66 sec
      Start 1539: test_gdasapp_util_smap2ioda
 8/44 Test #1539: test_gdasapp_util_smap2ioda ...........................   Passed    0.10 sec
      Start 1540: test_gdasapp_util_smos2ioda
 9/44 Test #1540: test_gdasapp_util_smos2ioda ...........................   Passed    0.10 sec
      Start 1541: test_gdasapp_util_viirsaod2ioda
10/44 Test #1541: test_gdasapp_util_viirsaod2ioda .......................   Passed    0.10 sec
      Start 1542: test_gdasapp_util_icecamsr2ioda
11/44 Test #1542: test_gdasapp_util_icecamsr2ioda .......................   Passed    0.11 sec
      Start 1543: test_gdasapp_util_icecmirs2ioda
12/44 Test #1543: test_gdasapp_util_icecmirs2ioda .......................   Passed    0.10 sec
      Start 1880: test_gdasapp_check_python_norms
13/44 Test #1880: test_gdasapp_check_python_norms .......................   Passed    3.87 sec
      Start 1881: test_gdasapp_check_yaml_keys
14/44 Test #1881: test_gdasapp_check_yaml_keys ..........................   Passed    1.54 sec
      Start 1882: test_gdasapp_jedi_increment_to_fv3
15/44 Test #1882: test_gdasapp_jedi_increment_to_fv3 ....................   Passed    8.02 sec
      Start 1883: test_gdasapp_fv3jedi_fv3inc
16/44 Test #1883: test_gdasapp_fv3jedi_fv3inc ...........................   Passed   23.24 sec
      Start 1884: test_gdasapp_soca_nsst_increment_to_mom6
17/44 Test #1884: test_gdasapp_soca_nsst_increment_to_mom6 ..............   Passed   34.90 sec
      Start 1885: test_gdasapp_soca_prep
18/44 Test #1885: test_gdasapp_soca_prep ................................   Passed    2.56 sec
      Start 1886: test_gdasapp_soca_run_clean
19/44 Test #1886: test_gdasapp_soca_run_clean ...........................   Passed    0.20 sec
      Start 1887: test_gdasapp_soca_setup_obsprep
20/44 Test #1887: test_gdasapp_soca_setup_obsprep .......................   Passed    9.72 sec
      Start 1888: test_gdasapp_soca_JGLOBAL_PREP_OCEAN_OBS
21/44 Test #1888: test_gdasapp_soca_JGLOBAL_PREP_OCEAN_OBS ..............   Passed  268.58 sec
      Start 1889: test_gdasapp_soca_JGLOBAL_MARINE_BMAT
22/44 Test #1889: test_gdasapp_soca_JGLOBAL_MARINE_BMAT .................   Passed  170.20 sec
      Start 1890: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP
23/44 Test #1890: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP ....   Passed   42.18 sec
      Start 1891: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_RUN
24/44 Test #1891: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_RUN .....   Passed   42.17 sec
      Start 1892: test_gdasapp_soca_copy_scratch
25/44 Test #1892: test_gdasapp_soca_copy_scratch ........................   Passed    0.20 sec
      Start 1893: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_CHKPT
26/44 Test #1893: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_CHKPT ...   Passed   42.19 sec
      Start 1894: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_POST
27/44 Test #1894: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_POST ....   Passed   42.17 sec
      Start 1895: test_gdasapp_snow_create_ens
28/44 Test #1895: test_gdasapp_snow_create_ens ..........................   Passed    1.01 sec
      Start 1896: test_gdasapp_snow_imsproc
29/44 Test #1896: test_gdasapp_snow_imsproc .............................   Passed    3.21 sec
      Start 1897: test_gdasapp_snow_apply_jediincr
30/44 Test #1897: test_gdasapp_snow_apply_jediincr ......................   Passed    3.25 sec
      Start 1898: test_gdasapp_snow_letkfoi_snowda
31/44 Test #1898: test_gdasapp_snow_letkfoi_snowda ......................   Passed    8.62 sec
      Start 1899: test_gdasapp_convert_bufr_adpsfc_snow
32/44 Test #1899: test_gdasapp_convert_bufr_adpsfc_snow .................   Passed    3.07 sec
      Start 1900: test_gdasapp_convert_bufr_adpsfc
33/44 Test #1900: test_gdasapp_convert_bufr_adpsfc ......................   Passed    2.99 sec
      Start 1901: test_gdasapp_convert_gsi_satbias
34/44 Test #1901: test_gdasapp_convert_gsi_satbias ......................   Passed    4.07 sec
      Start 1902: test_gdasapp_setup_atm_cycled_exp
35/44 Test #1902: test_gdasapp_setup_atm_cycled_exp .....................   Passed    0.60 sec
      Start 1903: test_gdasapp_atm_jjob_var_init
36/44 Test #1903: test_gdasapp_atm_jjob_var_init ........................   Passed  204.92 sec
      Start 1904: test_gdasapp_atm_jjob_var_run
37/44 Test #1904: test_gdasapp_atm_jjob_var_run .........................   Passed  106.22 sec
      Start 1905: test_gdasapp_atm_jjob_var_inc
38/44 Test #1905: test_gdasapp_atm_jjob_var_inc .........................   Passed   42.18 sec
      Start 1906: test_gdasapp_atm_jjob_var_final
39/44 Test #1906: test_gdasapp_atm_jjob_var_final .......................   Passed   42.15 sec
      Start 1907: test_gdasapp_atm_jjob_ens_init
40/44 Test #1907: test_gdasapp_atm_jjob_ens_init ........................   Passed   44.77 sec
      Start 1908: test_gdasapp_atm_jjob_ens_run
41/44 Test #1908: test_gdasapp_atm_jjob_ens_run .........................   Passed  490.23 sec
      Start 1909: test_gdasapp_atm_jjob_ens_inc
42/44 Test #1909: test_gdasapp_atm_jjob_ens_inc .........................   Passed  170.21 sec
      Start 1910: test_gdasapp_atm_jjob_ens_final
43/44 Test #1910: test_gdasapp_atm_jjob_ens_final .......................   Passed  234.21 sec
      Start 1911: test_gdasapp_aero_gen_3dvar_yaml
44/44 Test #1911: test_gdasapp_aero_gen_3dvar_yaml ......................   Passed    0.41 sec

100% tests passed, 0 tests failed out of 44

Label Time Summary:
gdas-utils    =  24.85 sec*proc (12 tests)
script        =  24.85 sec*proc (12 tests)

Total Test time (real) = 2079.10 sec
```

**Comment by RussTreadon-NOAA** [comment]
Automated testing on Hera and manual testing on Hercules yield _Passed_ for ctests.  The changes in this PR have been cycled in parallels on Hera and Hercules.  This PR is ready for review.

**Comment by RussTreadon-NOAA** [comment]
@emilyhcliu , are there any UFO validation developers you think should review this PR?

**Review by emilyhcliu — DISMISSED** [review]

**Comment by RussTreadon-NOAA** [comment]
Thank you @emilyhcliu

**Review by XuanliLi-NOAA — DISMISSED** [review]
Changes look good to me.

**Comment by RussTreadon-NOAA** [comment]
Thank you @XuanliLi-NOAA

**Review by emilyhcliu — DISMISSED** [review]

**Comment by XuanliLi-NOAA** [comment]
@RussTreadon-NOAA, with your new changes, the json file needs to be renamed to **bufr2ioda_gnssro.json** (currently  bufr2ioda_gnssro_bufr.json) and the bufr2ioda python script needs to be renamed to **bufr2ioda_gnssro.py** (currently bufr2ioda_gnssro_bufr.py).

**Comment by RussTreadon-NOAA** [comment]
@XuanliLi-NOAA , I made the requested change in a working copy of `feature/bufr2ioda` on Hercules.  prepatmiodaobs works fine with the name change.  

I have a question.  

`parm/ioda/bufr2ioda` and `ush/ioda/bufr2ioda` contains the following pair of files
- parm: `bufr2ioda_gnssro.json` and `bufr2ioda_gpsro_bufr_combined.json`
- ush: `bufr2ioda_gnssro.py` and `bufr2ioda_gpsro_bufr_combined.py`

Both python scripts read the same input `bufrfile`, namely `${cdump}.t${cyc}z.gpsro.tm00.bufr_d`, and write to different output `iodafile`
- `bufr2ioda_gnssro.py` creates `${cdump}.t${cyc}z.gnssro.tm00.nc`
- `bufr2ioda_gpsro_bufr_combined.py` creates `${cdump}.t${cyc}z..gpsro.tm00.nc`

I see both `gnssro.tm00.nc` and `gpsro.tm00.nc` created when I run prepatmiodaobs.   

However, there is no observation yaml for gpsro.  The [jcb-gdas](https://github.com/NOAA-EMC/jcb-gdas) repo only contains `gnssro.yaml.j2` in [`observations/atmosphere`](https://github.com/NOAA-EMC/jcb-gdas/blob/develop/observations/atmosphere) and [`observations/atmosphere-lgetk`](https://github.com/NOAA-EMC/jcb-gdas/blob/develop/observations/atmosphere-lgetkf).  There are no jcb-gdas yamls for gpsro in either `observations/atmosphere` or `observations/atmosphere-lgetkf`.

Do we need the gpsro bufr2ioda converter and json file?  
- If _yes_, it seems we should add gpsro yamls to jcb-gdas.
- If _no_, it seems we should remove the gpsro bufr2ioda converter and json file.

What do you think?

**Comment by XuanliLi-NOAA** [comment]
bufr2ioda_gpsro_bufr_combined.json and bufr2ioda_gpsro_bufr_combined.py are old files. We no longer need them.

**Comment by RussTreadon-NOAA** [comment]
Thank @XuanliLi-NOAA for letting me know that the two files in question are no longer needed.  Both files have been removed.  Done at 8d842b2.

**Comment by XuanliLi-NOAA** [comment]
Thank you!

File: parm/jcb-gdas
Comment (PraveenKumar-NOAA) [inline parm/jcb-gdas:1]: Change name for these two yamls as well: 
mtiasi_metop-a.yaml ----> mtiasi_metop-a.yaml.j2
mtiasi_metop-b.yaml ----> mtiasi_metop-b.yaml.j2

**Comment by RussTreadon-NOAA** [comment]
@emilyhcliu , @XuanliLi-NOAA , and @PraveenKumar-NOAA :  do you have any more comments or requests for this PR?  If not, would you please review and approve.  We would like to merge this PR into `develop`.

**Review by PraveenKumar-NOAA — APPROVED** [review]
Looks good!

File: parm/jcb-gdas
Comment (RussTreadon-NOAA) [reply parm/jcb-gdas:1]: Done at [jcb-gdas @ a132852](https://github.com/noaa-emc/jcb-gdas/tree/a132852df41b0b096d61b198f164c4c6e394662d)

Comment (PraveenKumar-NOAA) [reply parm/jcb-gdas:1]: @RussTreadon-NOAA thanks!

**Review by XuanliLi-NOAA — APPROVED** [review]
Looks good to me.

**Comment by RussTreadon-NOAA** [comment]
Thank you @XuanliLi-NOAA

**Comment by RussTreadon-NOAA** [comment]
Thank you @CoryMartin-NOAA .  I was just looking at this PR and the status changed.  Wow!

**Pull request merged**
